cli/integrity/datacollector.go

```diff
@@ -0,0 +1,35 @@
+package integrity
+
+import (
+	"time"
+
+	"github.com/tarantool/go-tarantool"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// Data represents collected data with its source.
+type Data struct {
+	// Source is the origin of data, i.e. key in case of etcd or tarantool-based collectors.
+	Source string
+	// Value is data collected.
+	Value []byte
+}
+
+// DataCollector interface must be implemented by a source collector.
+type DataCollector interface {
```

Comment:
What is the purpose of these types in the integrity package? Looks like they are not used here.

Reply:
Got it. I don't like that a separate package/modules makes assumptions about and depends on a higher-level functionality. I see why this is done like this. But all these collectors, publishers, they are not security package stuff at all and it makes this module API confusing, because it exports weird data types with unknown purpose for those, who did not read the `cluster` package code :) I need to think about it a little longer.

Reply:
This types will be moved to `lib/cluster` soon and everything will be nicer.

Reply:
`tt <- lib/integrity <- lib/cluster`
Does it mean `integrity` depends on `cluster`?

Reply:
Yes, it will depend.

But at now if we will move `integrity.DataCollector` into `cluster.DataCollector` it will be a circular dependency:
```
tt <- integrity <-> tt/cli/cluster
```
As result it will not compile.

But soon `tt/cli/cluster` will became a separate library, and we could move `integrity.DataCollector` to `cluster.DataCollector` (the same with other types which are more suitable for `cluster`) without circular dependency:
```
tt <- lib/integrity <- lib/cluster
tt <- lib/cluster
```
And everything will fine.

Reply:
> But at now if we will move integrity.DataCollector into cluster.DataCollector it will be a circular dependency:

It is a circular dependency due to the way the code is written. Because we move higher level collection-related code to the `integrity` package. The problem is not in `DataCollector` in the `integrity` package, the problem is that `integrity` package knows about data collecting and factories stuff, independently where the it is located. From my side of view it is an issue if the `integrity` package works with `DataCollector` or `cluster.DataCollector` or `SomeOtherModule.DataCollector` types.
Data collecting and publishing code should depend on `integrity` functionality and not the way around. `integrity` package/modules should provide an API to sign/verify files or messages (`[]byte`). It should not know something about etcd, tarantool or anything else. So whenever I want to sign something new, I don't need to change factory interface in separate `integrity` module. Just imagine, we decided to remove ETCD support completely. Should `integrity` module be updated? If yes, why? Is this expected?

> And everything will fine.

Looks like, I don't understand upcoming changes completely... If the `integrity` module will still use `cluster.*` types and the new `cluster` module will required `DataCollector` factory from the `integrity`, suppose it is still a circular dependency but between `cluster` and `integrity`.

Reply:
Right now, the trouble with moving `DataCollector` interface to `cluster` is that `integrity` is that we would want the following setup: `integrity` depends on `cluster`, and the "rest" of `tt` depends on both `cluster` and `integrity`. However, because we cannot import a single package from a module, as long as `cluster` package is part of `tt` module, depending on `cluster` package implies dependency on `tt` module. Therefore, we get circular dependency of `tt` and `integrity` modules. After `cluster` is moved to a separate module as well, depending on it will not imply dependency on `tt` module, allowing `integrity` to depend on `cluster`. Therefore, we will be able to move `DataCollector` and `DataPublisher` back to `cluster` and allow `integrity` dependency of `tt` on `cluster`.

As for the `integrity` not being an appropriate place for factories and collector/publisher instances -- I guess that's the price we pay for trying to hide the inner workings of integrity checking.

Reply:
`After cluster is moved to a separate module as well, depending on it will not imply dependency on tt module, allowing integrity to depend on cluster`

This is what I'm talking about. Why does `integrity` depend on `cluster`?

`As for the integrity not being an appropriate place for factories and collector/publisher instances -- I guess that's the price we pay for trying to hide the inner workings of integrity checking.`

So, cannot we move integrity-aware code to the `cluster` package with the verifying functionality still implemented somewhere else?

Reply:
> So, cannot we move integrity-aware code to the cluster package with the verifying functionality still implemented somewhere else?

This is somewhat tricky. We have some private fields in EE version of integrity, which are not exposed to avoid differences in API between CE and EE, that are required to perform integrity checks in `DataCollector`/`DataPublisher`

Reply:
Sorry, I don't understand how private fields in integrity implementation affect `cluster` package code.

Reply:
tt-ee has its own implementation of `DataCollector`/`DataPublisher` in `integrity` module that are used to collect/publish configurations with integrity checks.

Reply:
Still don't get it: why publishing and not only signing implementation?